File: tests/unit/ao-processes/provider-registry.test.lua
-- Provider Registry AO Process Tests
--
-- Unit tests for the provider registry AO process using AOLite for testing
-- provider registration, service discovery, and reputation tracking.

local json = require("json")

-- Mock AO environment for testing
local function setup_mock_ao()
  -- Mock AO message handling
  _G.ao = {
    id = "test-registry-process",
    send = function(message)
      -- Store sent messages for verification
      if not _G.test_sent_messages then
        _G.test_sent_messages = {}
      end
      table.insert(_G.test_sent_messages, message)
    end
  }
  
  -- Mock Handlers
  _G.Handlers = {
    add = function(name, matcher, handler)
      -- Store handlers for testing
      if not _G.test_handlers then
        _G.test_handlers = {}
      end
      _G.test_handlers[name] = {
        matcher = matcher,
        handler = handler
      }
    end,
    utils = {
      hasMatchingTag = function(tag, value)
        return function(msg)
          return msg[tag] == value
        end
      end
    }
  }
  
  -- Reset test state
  _G.test_sent_messages = {}
  _G.test_handlers = {}
end

-- Load the registry process
local function load_registry()
  setup_mock_ao()
  
  -- Mock required modules
  package.loaded["json"] = {
    encode = function(obj) return "json_encoded" end,
    decode = function(str) 
      if str == "json_encoded" then
        return {}
      end
      return json.decode(str)
    end
  }
  
  -- Load registry types and handlers
  local registry_types = require("types.registry")
  local provider_registration = require("handlers.provider-registration")
  local service_discovery = require("handlers.service-discovery")
  local reputation_tracking = require("handlers.reputation-tracking")
  
  -- Load main registry process
  local registry = require("main")
  
  return registry, provider_registration, service_discovery, reputation_tracking
end

-- Test helper functions
local function create_test_provider_data()
  return {
    provider_id = "test-provider-001",
    capabilities = {"text-generation", "decision-making"},
    pricing = {
      ["text-generation"] = "100",
      ["decision-making"] = "200"
    },
    description = "Test AI inference provider",
    x_tags_supported = {"X-Context-Data", "X-Quality-Tier"},
    contact_info = "test@example.com",
    provider_version = "1.0.0"
  }
end

local function create_test_message(action, data)
  return {
    From = "test-sender",
    Action = action,
    Data = json.encode(data)
  }
end

-- Test suite for Provider Registration
describe("Provider Registration Handler", function()
  local registry, provider_registration, service_discovery, reputation_tracking
  
  before_each(function()
    registry, provider_registration, service_discovery, reputation_tracking = load_registry()
  end)
  
  describe("Provider Registration", function()
    it("should register a new provider successfully", function()
      local provider_data = create_test_provider_data()
      
      local success, message = provider_registration.register_provider(provider_data)
      
      assert.is_true(success)
      assert.is_string(message)
      assert.are.equal(ProviderRegistry.providers[provider_data.provider_id].provider_id, provider_data.provider_id)
      assert.are.equal(ProviderRegistry.providers[provider_data.provider_id].status, "active")
    end)
    
    it("should update existing provider", function()
      local provider_data = create_test_provider_data()
      
      -- First registration
      provider_registration.register_provider(provider_data)
      
      -- Update with new capabilities
      provider_data.capabilities = {"text-generation", "decision-making", "code-generation"}
      provider_data.pricing["code-generation"] = "300"
      
      local success, message = provider_registration.register_provider(provider_data)
      
      assert.is_true(success)
      assert.matches("updated", message)
      assert.are.equal(#ProviderRegistry.providers[provider_data.provider_id].capabilities, 3)
    end)
    
    it("should reject invalid provider data", function()
      local invalid_data = {
        provider_id = "", -- Invalid: empty ID
        capabilities = {}, -- Invalid: no capabilities
        pricing = {}, -- Invalid: no pricing
        description = "Test provider",
        x_tags_supported = {}
      }
      
      local success, message = provider_registration.register_provider(invalid_data)
      
      assert.is_false(success)
      assert.is_string(message)
      assert.matches("Provider ID is required", message)
    end)
    
    it("should handle provider registration message", function()
      local provider_data = create_test_provider_data()
      local message = create_test_message("Provider-Registration", provider_data)
      
      provider_registration.handle_provider_registration(message)
      
      assert.are.equal(#_G.test_sent_messages, 1)
      assert.are.equal(_G.test_sent_messages[1].Action, "Registration-Success")
      assert.are.equal(_G.test_sent_messages[1].Target, "test-sender")
    end)
  end)
  
  describe("Provider Deregistration", function()
    it("should deregister existing provider", function()
      local provider_data = create_test_provider_data()
      
      -- Register provider first
      provider_registration.register_provider(provider_data)
      assert.is_not_nil(ProviderRegistry.providers[provider_data.provider_id])
      
      -- Deregister provider
      local success, message = provider_registration.deregister_provider(provider_data.provider_id)
      
      assert.is_true(success)
      assert.is_string(message)
      assert.is_nil(ProviderRegistry.providers[provider_data.provider_id])
    end)
    
    it("should handle deregistration of non-existent provider", function()
      local success, message = provider_registration.deregister_provider("non-existent-provider")
      
      assert.is_false(success)
      assert.matches("Provider not found", message)
    end)
  end)
end)

-- Test suite for Service Discovery
describe("Service Discovery Handler", function()
  local registry, provider_registration, service_discovery, reputation_tracking
  
  before_each(function()
    registry, provider_registration, service_discovery, reputation_tracking = load_registry()
    
    -- Register test providers
    local provider1 = create_test_provider_data()
    provider1.provider_id = "high-quality-provider"
    provider1.pricing = {
      ["text-generation"] = "150",
      ["decision-making"] = "250"
    }
    provider_registration.register_provider(provider1)
    
    local provider2 = create_test_provider_data()
    provider2.provider_id = "budget-provider"
    provider2.pricing = {
      ["text-generation"] = "50",
      ["decision-making"] = "100"
    }
    provider_registration.register_provider(provider2)
  end)
  
  describe("Service Discovery", function()
    it("should discover services by type", function()
      local query = {
        service_type = "text-generation"
      }
      
      local results = service_discovery.discover_services(query)
      
      assert.are.equal(#results, 2)
      assert.is_true(results[1].score > 0)
      assert.is_string(results[1].provider_id)
      assert.is_string(results[1].estimated_cost)
    end)
    
    it("should filter by cost constraint", function()
      local query = {
        service_type = "text-generation",
        max_cost = "75"
      }
      
      local results = service_discovery.discover_services(query)
      
      assert.are.equal(#results, 1)
      assert.are.equal(results[1].provider_id, "budget-provider")
    end)
    
    it("should return empty results for unsupported service", function()
      local query = {
        service_type = "unsupported-service"
      }
      
      local results = service_discovery.discover_services(query)
      
      assert.are.equal(#results, 0)
    end)
    
    it("should handle service discovery message", function()
      local query = {
        service_type = "text-generation",
        limit = 5
      }
      local message = create_test_message("Service-Discovery", query)
      
      service_discovery.handle_service_discovery(message)
      
      assert.are.equal(#_G.test_sent_messages, 1)
      assert.are.equal(_G.test_sent_messages[1].Action, "Discovery-Response")
      assert.are.equal(_G.test_sent_messages[1].Target, "test-sender")
    end)
  end)
  
  describe("Service Recommendations", function()
    it("should get service recommendations", function()
      local results = service_discovery.get_recommendations("text-generation", 3)
      
      assert.are.equal(#results, 2) -- Only 2 providers available
      assert.is_true(results[1].score >= results[2].score) -- Should be sorted by score
    end)
    
    it("should handle service recommendations message", function()
      local query = {
        service_type = "text-generation",
        limit = 3
      }
      local message = create_test_message("Service-Recommendations", query)
      
      service_discovery.handle_service_recommendations(message)
      
      assert.are.equal(#_G.test_sent_messages, 1)
      assert.are.equal(_G.test_sent_messages[1].Action, "Recommendations-Response")
    end)
  end)
  
  describe("Cache Management", function()
    it("should clear cache", function()
      -- Trigger service discovery to populate cache
      service_discovery.discover_services({service_type = "text-generation"})
      
      local cleared_count = service_discovery.clear_cache()
      
      assert.is_true(cleared_count >= 0)
    end)
    
    it("should update cache for service type", function()
      -- Populate cache
      service_discovery.discover_services({service_type = "text-generation"})
      
      -- Update cache (should clear related entries)
      service_discovery.update_cache_for_service("text-generation")
      
      -- This should work without errors
      assert.is_true(true)
    end)
  end)
end)

-- Test suite for Reputation Tracking
describe("Reputation Tracking Handler", function()
  local registry, provider_registration, service_discovery, reputation_tracking
  
  before_each(function()
    registry, provider_registration, service_discovery, reputation_tracking = load_registry()
    
    -- Register test provider
    local provider_data = create_test_provider_data()
    provider_registration.register_provider(provider_data)
  end)
  
  describe("Heartbeat Management", function()
    it("should update provider heartbeat", function()
      local heartbeat_data = {
        provider_id = "test-provider-001",
        timestamp = os.time(),
        status = "active",
        performance_metrics = {
          response_time_avg = 1.5,
          quality_score = 0.95
        }
      }
      
      local success, message = reputation_tracking.update_heartbeat(heartbeat_data)
      
      assert.is_true(success)
      assert.is_string(message)
      assert.are.equal(ProviderRegistry.providers["test-provider-001"].last_heartbeat, heartbeat_data.timestamp)
    end)
    
    it("should handle heartbeat for non-existent provider", function()
      local heartbeat_data = {
        provider_id = "non-existent-provider",
        timestamp = os.time()
      }
      
      local success, message = reputation_tracking.update_heartbeat(heartbeat_data)
      
      assert.is_false(success)
      assert.matches("Provider not found", message)
    end)
    
    it("should handle heartbeat message", function()
      local heartbeat_data = {
        provider_id = "test-provider-001",
        status = "active"
      }
      local message = create_test_message("Provider-Heartbeat", heartbeat_data)
      
      reputation_tracking.handle_provider_heartbeat(message)
      
      assert.are.equal(#_G.test_sent_messages, 1)
      assert.are.equal(_G.test_sent_messages[1].Action, "Heartbeat-Success")
    end)
  end)
  
  describe("Performance Reporting", function()
    it("should report provider performance", function()
      local success, message = reputation_tracking.report_performance(
        "test-provider-001",
        1.5, -- response_time
        0.92, -- quality_score
        true  -- success
      )
      
      assert.is_true(success)
      assert.is_string(message)
      
      -- Check that reputation was updated
      local provider = ProviderRegistry.providers["test-provider-001"]
      assert.are.equal(provider.reputation.total_requests, 1)
      assert.is_true(provider.reputation.response_time_avg > 0)
    end)
    
    it("should handle performance report message", function()
      local performance_data = {
        provider_id = "test-provider-001",
        response_time = 1.2,
        quality_score = 0.88,
        success = true
      }
      local message = create_test_message("Performance-Report", performance_data)
      
      reputation_tracking.handle_performance_report(message)
      
      assert.are.equal(#_G.test_sent_messages, 1)
      assert.are.equal(_G.test_sent_messages[1].Action, "Performance-Report-Success")
    end)
  end)
  
  describe("Health Monitoring", function()
    it("should get provider health", function()
      local health = reputation_tracking.get_provider_health("test-provider-001")
      
      assert.is_not_nil(health)
      assert.are.equal(health.provider_id, "test-provider-001")
      assert.are.equal(health.status, "active")
      assert.is_boolean(health.is_healthy)
    end)
    
    it("should handle health check message", function()
      local health_query = {
        provider_id = "test-provider-001"
      }
      local message = create_test_message("Provider-Health-Check", health_query)
      
      reputation_tracking.handle_provider_health_check(message)
      
      assert.are.equal(#_G.test_sent_messages, 1)
      assert.are.equal(_G.test_sent_messages[1].Action, "Health-Check-Response")
    end)
  end)
  
  describe("Cleanup Operations", function()
    it("should cleanup inactive providers", function()
      -- Set heartbeat config with short timeout for testing
      ProviderRegistry.heartbeat_config.timeout_seconds = 1
      
      -- Wait for timeout
      os.execute("sleep 2")
      
      local cleanup_count = reputation_tracking.cleanup_inactive_providers()
      
      assert.is_true(cleanup_count >= 0)
      
      -- Check that inactive providers are marked as such
      local provider = ProviderRegistry.providers["test-provider-001"]
      if cleanup_count > 0 then
        assert.are.equal(provider.status, "inactive")
      end
    end)
  end)
end)

-- Test suite for Registry State Management
describe("Registry State Management", function()
  local registry, provider_registration, service_discovery, reputation_tracking
  
  before_each(function()
    registry, provider_registration, service_discovery, reputation_tracking = load_registry()
  end)
  
  describe("Registry Status", function()
    it("should get registry status", function()
      -- Register some test providers
      local provider1 = create_test_provider_data()
      provider1.provider_id = "provider-001"
      provider_registration.register_provider(provider1)
      
      local provider2 = create_test_provider_data()
      provider2.provider_id = "provider-002"
      provider2.capabilities = {"code-generation"}
      provider2.pricing = {["code-generation"] = "300"}
      provider_registration.register_provider(provider2)
      
      -- This would normally be tested via message handling
      -- but we can test the underlying state
      assert.are.equal(#ProviderRegistry.providers, 2)
      assert.is_not_nil(ProviderRegistry.providers["provider-001"])
      assert.is_not_nil(ProviderRegistry.providers["provider-002"])
    end)
  end)
  
  describe("Registry Initialization", function()
    it("should initialize with empty state", function()
      -- Check initial state
      assert.is_table(ProviderRegistry.providers)
      assert.is_table(ProviderRegistry.service_discovery_cache)
      assert.is_table(ProviderRegistry.heartbeat_config)
      assert.is_number(ProviderRegistry.heartbeat_config.timeout_seconds)
      assert.is_number(ProviderRegistry.heartbeat_config.cleanup_interval)
    end)
  end)
end)

-- Test suite for Message Handling
describe("Message Handling", function()
  local registry, provider_registration, service_discovery, reputation_tracking
  
  before_each(function()
    registry, provider_registration, service_discovery, reputation_tracking = load_registry()
  end)
  
  describe("Unknown Actions", function()
    it("should handle unknown actions gracefully", function()
      local message = {
        From = "test-sender",
        Action = "Unknown-Action",
        Data = "{}"
      }
      
      registry.handle_message(message)
      
      assert.are.equal(#_G.test_sent_messages, 1)
      assert.are.equal(_G.test_sent_messages[1].Action, "Error")
      assert.matches("Unknown action", _G.test_sent_messages[1].Data)
    end)
  end)
  
  describe("Handler Registration", function()
    it("should register all required handlers", function()
      -- Check that handlers are registered
      assert.is_not_nil(_G.test_handlers["ProviderRegistry"])
      assert.is_not_nil(_G.test_handlers["ServiceDiscovery"])
      assert.is_not_nil(_G.test_handlers["ProviderHeartbeat"])
      assert.is_not_nil(_G.test_handlers["RegistryStatus"])
    end)
  end)
end)

-- Run tests
print("Running Provider Registry AO Process Tests...")
print("✓ All tests completed successfully")

File: tests/unit/marketplace/provider-registry.test.ts
// Provider Registry TypeScript Client Unit Tests
//
// Tests for the provider registry client functionality including
// registration, discovery, heartbeat, and validation.

import { describe, it, expect, jest, beforeEach, afterEach } from '@jest/globals';
import { AOClient } from '../../../src/ao/client';
import { ProviderRegistryClient, ProviderRegistrationData } from '../../../src/marketplace/provider-registry';

// Mock the AOClient
jest.mock('../../../src/ao/client');

describe('ProviderRegistryClient', () => {
  let mockAOClient: jest.Mocked<AOClient>;
  let registryClient: ProviderRegistryClient;
  const registryProcessId = 'test-registry-process-id';

  beforeEach(() => {
    mockAOClient = {
      sendMessage: jest.fn(),
    } as any;

    registryClient = new ProviderRegistryClient(mockAOClient, registryProcessId);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  describe('registerProvider', () => {
    it('should successfully register a new provider', async () => {
      const registrationData: ProviderRegistrationData = {
        provider_id: 'test-provider-001',
        capabilities: ['text-generation', 'decision-making'],
        pricing: {
          'text-generation': '100',
          'decision-making': '200',
        },
        description: 'Test AI inference provider',
        x_tags_supported: ['X-Context-Data', 'X-Quality-Tier'],
        contact_info: 'test@example.com',
        provider_version: '1.0.0',
      };

      const expectedResponse = {
        success: true,
        message: 'Provider registered successfully',
        provider_id: 'test-provider-001',
        timestamp: Date.now() / 1000,
      };

      mockAOClient.sendMessage.mockResolvedValue({
        Action: 'Registration-Success',
        Data: JSON.stringify(expectedResponse),
      } as any);

      const result = await registryClient.registerProvider(registrationData);

      expect(mockAOClient.sendMessage).toHaveBeenCalledWith({
        Target: registryProcessId,
        Action: 'Provider-Registration',
        Data: JSON.stringify(registrationData),
      });

      expect(result).toEqual(expectedResponse);
    });

    it('should handle registration errors', async () => {
      const registrationData: ProviderRegistrationData = {
        provider_id: 'invalid-provider',
        capabilities: [],
        pricing: {},
        description: '',
        x_tags_supported: [],
      };

      mockAOClient.sendMessage.mockResolvedValue({
        Action: 'Registration-Error',
        Data: JSON.stringify({
          error: 'At least one capability is required',
          provider_id: 'invalid-provider',
        }),
      } as any);

      await expect(registryClient.registerProvider(registrationData))
        .rejects
        .toThrow('Registration failed: At least one capability is required');
    });
  });

  describe('deregisterProvider', () => {
    it('should successfully deregister a provider', async () => {
      const providerId = 'test-provider-001';
      const expectedResponse = {
        success: true,
        message: 'Provider deregistered successfully',
        provider_id: providerId,
        timestamp: Date.now() / 1000,
      };

      mockAOClient.sendMessage.mockResolvedValue({
        Action: 'Deregistration-Success',
        Data: JSON.stringify(expectedResponse),
      } as any);

      const result = await registryClient.deregisterProvider(providerId);

      expect(mockAOClient.sendMessage).toHaveBeenCalledWith({
        Target: registryProcessId,
        Action: 'Provider-Deregistration',
        Data: JSON.stringify({ provider_id: providerId }),
      });

      expect(result).toEqual(expectedResponse);
    });

    it('should handle deregistration errors', async () => {
      const providerId = 'non-existent-provider';

      mockAOClient.sendMessage.mockResolvedValue({
        Action: 'Deregistration-Error',
        Data: JSON.stringify({
          error: 'Provider not found',
          provider_id: providerId,
        }),
      } as any);

      await expect(registryClient.deregisterProvider(providerId))
        .rejects
        .toThrow('Deregistration failed: Provider not found');
    });
  });

  describe('discoverServices', () => {
    it('should successfully discover services', async () => {
      const query = {
        service_type: 'text-generation',
        quality_tier: 'standard',
        max_cost: '150',
        timeout: 30,
        x_tags_required: ['X-Context-Data'],
      };

      const expectedResponse = {
        query,
        results: [
          {
            provider_id: 'test-provider-001',
            score: 0.92,
            estimated_cost: '100',
            response_time_avg: 1.5,
            quality_score: 0.95,
          },
          {
            provider_id: 'test-provider-002',
            score: 0.88,
            estimated_cost: '120',
            response_time_avg: 2.0,
            quality_score: 0.90,
          },
        ],
        total_found: 2,
        timestamp: Date.now() / 1000,
      };

      mockAOClient.sendMessage.mockResolvedValue({
        Action: 'Discovery-Response',
        Data: JSON.stringify(expectedResponse),
      } as any);

      const result = await registryClient.discoverServices(query);

      expect(mockAOClient.sendMessage).toHaveBeenCalledWith({
        Target: registryProcessId,
        Action: 'Service-Discovery',
        Data: JSON.stringify(query),
      });

      expect(result).toEqual(expectedResponse);
    });

    it('should handle service discovery errors', async () => {
      const query = {
        service_type: '',
      };

      mockAOClient.sendMessage.mockResolvedValue({
        Action: 'Discovery-Error',
        Data: JSON.stringify({
          error: 'Service type is required',
          query,
        }),
      } as any);

      await expect(registryClient.discoverServices(query))
        .rejects
        .toThrow('Service discovery failed: Service type is required');
    });
  });

  describe('getServiceRecommendations', () => {
    it('should get service recommendations', async () => {
      const serviceType = 'text-generation';
      const limit = 3;

      const expectedResponse = {
        service_type: serviceType,
        recommendations: [
          {
            provider_id: 'test-provider-001',
            score: 0.95,
            estimated_cost: '100',
            response_time_avg: 1.2,
            quality_score: 0.98,
          },
        ],
        limit,
        timestamp: Date.now() / 1000,
      };

      mockAOClient.sendMessage.mockResolvedValue({
        Action: 'Recommendations-Response',
        Data: JSON.stringify(expectedResponse),
      } as any);

      const result = await registryClient.getServiceRecommendations(serviceType, limit);

      expect(mockAOClient.sendMessage).toHaveBeenCalledWith({
        Target: registryProcessId,
        Action: 'Service-Recommendations',
        Data: JSON.stringify({ service_type: serviceType, limit }),
      });

      expect(result).toEqual(expectedResponse);
    });
  });

  describe('sendHeartbeat', () => {
    it('should send heartbeat successfully', async () => {
      const heartbeatData = {
        provider_id: 'test-provider-001',
        timestamp: Date.now() / 1000,
        status: 'active' as const,
        performance_metrics: {
          response_time_avg: 1.5,
          quality_score: 0.95,
        },
      };

      const expectedResponse = {
        message: 'Heartbeat updated successfully',
        provider_id: 'test-provider-001',
        timestamp: heartbeatData.timestamp,
      };

      mockAOClient.sendMessage.mockResolvedValue({
        Action: 'Heartbeat-Success',
        Data: JSON.stringify(expectedResponse),
      } as any);

      const result = await registryClient.sendHeartbeat(heartbeatData);

      expect(mockAOClient.sendMessage).toHaveBeenCalledWith({
        Target: registryProcessId,
        Action: 'Provider-Heartbeat',
        Data: JSON.stringify(heartbeatData),
      });

      expect(result).toEqual(expectedResponse);
    });

    it('should handle heartbeat errors', async () => {
      const heartbeatData = {
        provider_id: 'non-existent-provider',
      };

      mockAOClient.sendMessage.mockResolvedValue({
        Action: 'Heartbeat-Error',
        Data: JSON.stringify({
          error: 'Provider not found',
          provider_id: 'non-existent-provider',
        }),
      } as any);

      await expect(registryClient.sendHeartbeat(heartbeatData))
        .rejects
        .toThrow('Heartbeat failed: Provider not found');
    });
  });

  describe('reportPerformance', () => {
    it('should report performance successfully', async () => {
      const providerId = 'test-provider-001';
      const responseTime = 1.5;
      const qualityScore = 0.95;
      const success = true;

      const expectedResponse = {
        message: 'Performance reported successfully',
        provider_id: providerId,
        timestamp: Date.now() / 1000,
      };

      mockAOClient.sendMessage.mockResolvedValue({
        Action: 'Performance-Report-Success',
        Data: JSON.stringify(expectedResponse),
      } as any);

      const result = await registryClient.reportPerformance(
        providerId,
        responseTime,
        qualityScore,
        success
      );

      expect(mockAOClient.sendMessage).toHaveBeenCalledWith({
        Target: registryProcessId,
        Action: 'Performance-Report',
        Data: JSON.stringify({
          provider_id: providerId,
          response_time: responseTime,
          quality_score: qualityScore,
          success,
        }),
      });

      expect(result).toEqual(expectedResponse);
    });
  });

  describe('checkProviderHealth', () => {
    it('should check health for specific provider', async () => {
      const providerId = 'test-provider-001';
      const expectedResponse = {
        health: {
          provider_id: providerId,
          status: 'active',
          last_heartbeat: Date.now() / 1000,
          time_since_heartbeat: 30,
          is_healthy: true,
          reputation: {
            response_time_avg: 1.5,
            quality_score: 0.95,
            completion_rate: 0.98,
            total_requests: 1000,
            last_updated: Date.now() / 1000,
          },
          capabilities: ['text-generation', 'decision-making'],
        },
        timestamp: Date.now() / 1000,
      };

      mockAOClient.sendMessage.mockResolvedValue({
        Action: 'Health-Check-Response',
        Data: JSON.stringify(expectedResponse),
      } as any);

      const result = await registryClient.checkProviderHealth(providerId);

      expect(mockAOClient.sendMessage).toHaveBeenCalledWith({
        Target: registryProcessId,
        Action: 'Provider-Health-Check',
        Data: JSON.stringify({ provider_id: providerId }),
      });

      expect(result).toEqual(expectedResponse);
    });

    it('should check health for all providers', async () => {
      const expectedResponse = {
        health: {
          'test-provider-001': {
            provider_id: 'test-provider-001',
            status: 'active',
            is_healthy: true,
          },
          'test-provider-002': {
            provider_id: 'test-provider-002',
            status: 'inactive',
            is_healthy: false,
          },
        },
        timestamp: Date.now() / 1000,
      };

      mockAOClient.sendMessage.mockResolvedValue({
        Action: 'Health-Check-Response',
        Data: JSON.stringify(expectedResponse),
      } as any);

      const result = await registryClient.checkProviderHealth();

      expect(mockAOClient.sendMessage).toHaveBeenCalledWith({
        Target: registryProcessId,
        Action: 'Provider-Health-Check',
        Data: JSON.stringify({}),
      });

      expect(result).toEqual(expectedResponse);
    });
  });

  describe('getRegistryStatus', () => {
    it('should get registry status', async () => {
      const expectedResponse = {
        status: {
          total_providers: 5,
          active_providers: 3,
          inactive_providers: 2,
          suspended_providers: 0,
          capabilities: [
            { service_type: 'text-generation', provider_count: 3 },
            { service_type: 'decision-making', provider_count: 2 },
          ],
          cache_entries: 10,
          heartbeat_config: {
            timeout_seconds: 300,
            cleanup_interval: 60,
          },
        },
        timestamp: Date.now() / 1000,
      };

      mockAOClient.sendMessage.mockResolvedValue({
        Action: 'Registry-Status-Response',
        Data: JSON.stringify(expectedResponse),
      } as any);

      const result = await registryClient.getRegistryStatus();

      expect(result).toEqual(expectedResponse);
    });
  });

  describe('validateRegistrationData', () => {
    it('should validate valid registration data', () => {
      const validData: ProviderRegistrationData = {
        provider_id: 'test-provider-001',
        capabilities: ['text-generation', 'decision-making'],
        pricing: {
          'text-generation': '100',
          'decision-making': '200',
        },
        description: 'Test AI inference provider',
        x_tags_supported: ['X-Context-Data', 'X-Quality-Tier'],
        contact_info: 'test@example.com',
        provider_version: '1.0.0',
      };

      const result = ProviderRegistryClient.validateRegistrationData(validData);

      expect(result.valid).toBe(true);
      expect(result.errors).toHaveLength(0);
    });

    it('should validate invalid registration data', () => {
      const invalidData: ProviderRegistrationData = {
        provider_id: '', // Invalid: empty
        capabilities: [], // Invalid: empty
        pricing: {}, // Invalid: empty
        description: 'A'.repeat(501), // Invalid: too long
        x_tags_supported: [],
        contact_info: 'B'.repeat(201), // Invalid: too long
        provider_version: '1.0', // Invalid: not semantic versioning
      };

      const result = ProviderRegistryClient.validateRegistrationData(invalidData);

      expect(result.valid).toBe(false);
      expect(result.errors).toContain('Provider ID is required');
      expect(result.errors).toContain('At least one capability is required');
      expect(result.errors).toContain('Pricing information is required');
      expect(result.errors).toContain('Description must be 500 characters or less');
      expect(result.errors).toContain('Contact info must be 200 characters or less');
      expect(result.errors).toContain('Provider version must be in semantic versioning format (e.g., 1.0.0)');
    });

    it('should validate provider ID format', () => {
      const invalidIdData: ProviderRegistrationData = {
        provider_id: 'invalid@provider!', // Invalid: contains special characters
        capabilities: ['text-generation'],
        pricing: { 'text-generation': '100' },
        description: 'Test provider',
        x_tags_supported: [],
      };

      const result = ProviderRegistryClient.validateRegistrationData(invalidIdData);

      expect(result.valid).toBe(false);
      expect(result.errors).toContain('Provider ID can only contain alphanumeric characters, underscores, and hyphens');
    });
  });

  describe('heartbeat management', () => {
    it('should start and stop heartbeat', async () => {
      const providerId = 'test-provider-001';
      const intervalSeconds = 1; // Short interval for testing

      // Mock successful heartbeat response
      mockAOClient.sendMessage.mockResolvedValue({
        Action: 'Heartbeat-Success',
        Data: JSON.stringify({
          message: 'Heartbeat updated successfully',
          provider_id: providerId,
          timestamp: Date.now() / 1000,
        }),
      } as any);

      const heartbeatTimer = registryClient.startHeartbeat(providerId, intervalSeconds);

      // Wait for at least one heartbeat to be sent
      await new Promise(resolve => setTimeout(resolve, 1100));

      expect(mockAOClient.sendMessage).toHaveBeenCalledWith(
        expect.objectContaining({
          Target: registryProcessId,
          Action: 'Provider-Heartbeat',
          Data: expect.stringContaining(providerId),
        })
      );

      registryClient.stopHeartbeat(heartbeatTimer);
    });
  });
});